src/artifact.rs

```diff
@@ -447,10 +447,36 @@ impl Artifact {
 
     }
 
+    /// Get set of non-import declarations that have not been defined. This must be an empty set in
+    /// order to `emit` the artifact.
+    pub fn undefined_symbols(&self) -> Vec<String> {
+        let mut decls = self.declarations.clone();
+        for def in self.definitions.iter() {
+            decls.remove(&def.name);
+        }
+
+        let mut syms = Vec::new();
+        for (&name, &decl) in decls.iter() {
+            match decl {
+                Decl::FunctionImport | Decl::DataImport => {},
+                Decl::Function { .. } | Decl::Data { .. } | Decl::CString { .. } => {
```

Comment:
I just noticed that `Decl` has an `is_import()` function, which would tidy things up here.